app/services/db/identity/iaa_active_user_count.rb

```diff
@@ -0,0 +1,97 @@
+# generate incremental IAL2 active user counts by IAA (rolling up applications)
+# incremental means a user can only be active for one month in the entire span of the IAA
+#
+# strategy: get active users in the current month in identities and
+# subtract out those active in prior months from monthly_sp_auth_counts
+#
+# for each IAA:
+# SELECT COUNT(*) as active_count FROM identities
+# WHERE service_provider in (issuer1, issuer2, ...)
+# AND last_ial2_authenticated_at >= first_of_the_month
+# AND user_id IN (select user_id from profiles)
+# AND user_id NOT IN (
+#   SELECT DISTINCT user_id
+#   FROM monthly_sp_auth_counts
+#   WHERE issuer IN (issuer1, issuer2, ...) and ial=2
+#   and year_month in (iaa_start_year_month, iaa_start_year_month+1, ... previous_month)
+
+module Db
+  module Identity
+    class IaaActiveUserCount
+      def initialize(iaa, iaa_start_date, iaa_end_date)
+        @iaa = iaa
+        @iaa_start_date = iaa_start_date
+        @iaa_end_date = iaa_end_date
+      end
+
+      def call(ial, today)
+        issuers = issuers_sql(iaa)
+        return unless issuers
+        prior_months = prior_months_sql(today)
+        prior_months = from_monthly_sp_counts(issuers, ial, prior_months) if prior_months.present?
+        sql = format(<<~SQL, sql_params(today))
+          SELECT COUNT(*) as active_count FROM identities
+          WHERE service_provider in #{issuers}
+          AND last_ial#{ial}_authenticated_at >= %{beginning_of_month} and
+          user_id IN (select user_id from profiles)
+          #{prior_months}
+        SQL
+        ActiveRecord::Base.connection.execute(sql)[0]['active_count']
+      end
+
+      private
+
+      attr_reader :iaa, :iaa_start_date, :iaa_end_date
+
+      def from_monthly_sp_counts(issuers, ial, prior_months)
+        <<~SQL
+          AND user_id NOT IN (
```

Comment:
This might be a bit goofy (and my understanding of the structure may be off), but a potential improvement in readability and database speed:

If `year_month` is in the format "YYYYMM", this could be turned into a `between` query since that format is sortable. Ex:

```ruby
"year_month BETWEEN '#{iaa_start_date.strftime("%Y%m")}' AND '#{today.strftime("%Y%m")}'"
# => "year_month BETWEEN '202004' AND '202008'"
```

Reply:
great idea!